service/mapper/transitions.go

```diff
@@ -0,0 +1,380 @@
+	// of payloads and just go straight to forwarding the height to the next
+	// finalized block.
+	if !t.cfg.IndexPayloads {
+		s.status = StatusIndexed
+		return nil
+	}
+
+	// If we index payloads, we are basically stepping back from (and including)
+	// the tree that corresponds to the next finalized block all the way up to
+	// (and excluding) the tree for the last finalized block we indexed. To do
+	// so, we will use the parent state commit to retrieve the parent trees from
+	// the forest, and we use the paths we recorded changes on to retrieve the
+	// changed payloads at each step.
+	commit := s.next
+	for commit != s.last {
+
+		// We do this check only once, so that we don't need to do it for
+		// each item we retrieve. The tree should always be there, but we
+		// should check just to not fail silently.
+		ok := s.forest.Has(commit)
+		if !ok {
+			return fmt.Errorf("could not load tree (commit: %x)", commit)
+		}
+
+		// For each path, we retrieve the payload and add it to the registers we
+		// will index later. If we already have a payload for the path, it is
+		// more recent as we iterate backwards in time, so we can skip the
+		// outdated payload.
+		// NOTE: We read from the tree one by one here, as the performance
+		// overhead is minimal compared to the disk i/o for badger, and it
+		// allows us to ignore sorting of paths.
+		tree, _ := s.forest.Tree(commit)
+		paths, _ := s.forest.Paths(commit)
+		for _, path := range paths {
+			_, ok := s.registers[path]
+			if ok {
+				continue
+			}
+			payloads := tree.UnsafeRead([]ledger.Path{path})
+			s.registers[path] = payloads[0]
+		}
+
+		// We now step back to the parent of the current state trie.
+		parent, _ := s.forest.Parent(commit)
+		commit = parent
+	}
+
+	log.Info().Int("registers", len(s.registers)).Msg("collected registers for finalized block")
+
+	// At this point, we have collected all the payloads, so we go to the next
+	// step, where we will index them.
+	s.status = StatusCollected
+
+	return nil
+}
+
+func (t *Transitions) IndexRegisters(s *State) error {
+	log := t.log.With().Uint64("height", s.height).Hex("commit", s.next[:]).Logger()
+
+	// We should only index the payloads if we have just collected the payloads
+	// of a finalized block.
+	if s.status != StatusCollected {
+		return fmt.Errorf("invalid status for indexing registers (%s)", s.status)
+	}
+
+	// If there are no registers left to be indexed, we can go to the next step,
+	// which is about forwarding the height to the next finalized block.
+	if len(s.registers) == 0 {
+		log.Info().Msg("indexed registers for finalized block")
+		s.status = StatusIndexed
+		return nil
+	}
+
+	// We will now collect and index 1000 registers at a time. This gives the
+	// FSM the chance to exit the loop between every 1000 payloads we index. It
+	// doesn't really matter for badger if they are in random order, so this
+	// way of iterating should be fine.
+	n := 1000
```

Comment:
We could make this a const so that the threshold value is a bit more visible.

Reply:
Nope, this should be a variable, not a constant. If anything, it should be made into a configurable variable.